cmake/Protobuf.cmake

```diff
@@ -0,0 +1,76 @@
+macro(p4c_obtain_protobuf)
+  option(
+    P4C_USE_PREINSTALLED_PROTOBUF
+    "Look for a preinstalled version of Protobuf in the system instead of installing a prebuilt binary using FetchContent."
+    OFF
+  )
+
+  # If P4C_USE_PREINSTALLED_PROTOBUF is ON just try to find a preinstalled version of Protobuf.
+  if(P4C_USE_PREINSTALLED_PROTOBUF)
+    if(ENABLE_PROTOBUF_STATIC)
+      set(SAVED_CMAKE_FIND_LIBRARY_SUFFIXES ${CMAKE_FIND_LIBRARY_SUFFIXES})
+      set(CMAKE_FIND_LIBRARY_SUFFIXES .a)
+    endif()
+    find_package(Protobuf 3.0.0 REQUIRED)
+    if(ENABLE_PROTOBUF_STATIC)
+      set(CMAKE_FIND_LIBRARY_SUFFIXES ${SAVED_CMAKE_FIND_LIBRARY_SUFFIXES})
+    endif()
+  else()
+    # Google introduced another breaking change with protobuf 22.x by adding abseil as a new
+    # dependency. https://protobuf.dev/news/2022-08-03/#abseil-dep We do not want abseil, so we stay
+    # with 21.x for now.
+    set(P4C_PROTOBUF_VERSION "21.12")
+    message("Fetching Protobuf version ${P4C_PROTOBUF_VERSION} for P4C...")
+
+    # Print out download state while setting up Protobuf.
+    set(FETCHCONTENT_QUIET_PREV ${FETCHCONTENT_QUIET})
+    set(FETCHCONTENT_QUIET OFF)
+    set(protobuf_BUILD_TESTS OFF CACHE BOOL "Build tests.")
+    set(protobuf_BUILD_PROTOC_BINARIES OFF CACHE BOOL "Build libprotoc and protoc compiler.")
+    # Only ever build the static library. It is not safe to link with a local dynamic version.
+    set(protobuf_BUILD_SHARED_LIBS OFF CACHE BOOL "Build Shared Libraries")
+    # Unity builds do not work for Protobuf...
+    set(SAVED_CMAKE_UNITY_BUILD ${CMAKE_UNITY_BUILD})
+    set(CMAKE_UNITY_BUILD OFF)
+    fetchcontent_declare(
+      protobuf
+      URL https://github.com/protocolbuffers/protobuf/releases/download/v${P4C_PROTOBUF_VERSION}/protobuf-cpp-3.${P4C_PROTOBUF_VERSION}.tar.gz
+      URL_HASH SHA256=4eab9b524aa5913c6fffb20b2a8abf5ef7f95a80bc0701f3a6dbb4c607f73460
+      USES_TERMINAL_DOWNLOAD TRUE
+      GIT_PROGRESS TRUE
+    )
+    # Pull a different protoc binary for MacOS.
+    # TODO: Should we build from scratch?
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
+      fetchcontent_declare(
+        protoc
+        URL https://github.com/protocolbuffers/protobuf/releases/download/v${P4C_PROTOBUF_VERSION}/protoc-${P4C_PROTOBUF_VERSION}-osx-x86_64.zip
+        USES_TERMINAL_DOWNLOAD TRUE
+        GIT_PROGRESS TRUE
+      )
+    else()
+      fetchcontent_declare(
+        protoc
+        URL https://github.com/protocolbuffers/protobuf/releases/download/v${P4C_PROTOBUF_VERSION}/protoc-${P4C_PROTOBUF_VERSION}-linux-x86_64.zip
```

Comment:
@fruffy This seems to break the build for architectures other than x86_64 (e.g., aarch64).

Reply:
Afaik we do not support or make guarantees for any architecture other than `x86_64`. But if there is a need this should be easily patchable.

Reply:
p4c is also used on ARM64 platforms like [P4EDGE](https://p4edge.net/) and [P4Pi](https://github.com/p4lang/p4pi). It would be great if we could also enable support for the `aarch64` architecture.

Reply:
Using the option `P4C_USE_PREINSTALLED_PROTOBUF=ON` will use the system's Protobuf version. In that case things should compile as usual until we have a patch for these types of architectures in place.

Reply:
Thank Fabian!

Some platforms are not connected to the internet and FetchContent would fail when building p4c. As a workaround, we can run `cmake` on a system with internet access and include `build/_deps` with the source code.

Do you have any advice on how to enable building p4c offline?

Reply:
I dont think that building without internet is a supported usecase.

Reply:
There are ways to run FetchContent fully disconnected https://cmake.org/cmake/help/latest/module/FetchContent.html#command:fetchcontent_populate
At least I have been in scenarios where I did not have internet and was able to build as usual.

Now, populating `build/deps` in the beginning is a different story. But that is no different that running `git submodule init --update`. I do not have a good suggestion there yet other than using the PREINSTALLED options.